lib/rbtree/rbtree.h

```diff
@@ -0,0 +1,455 @@
+#ifndef __INCLUDED_RBTREE_H__
+#define __INCLUDED_RBTREE_H__
+
+/** \file rbtree.h
+ * Declaration of associated structures and functions for a simple, intrusive
+ * red-black tree implementation.
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+#include <stdlib.h>
+#include <assert.h>
+
+/** \defgroup rb_tree_compiler_prims Compiler Abstractions
+ * Primitives used to abstract compiler-specific syntax for common details used in
+ * providing hints to the compiler for optimization or linker details.
+ * @{
+ */
+
+/**
+ * Macro to check if a given assertion about an argument is true
+ */
+#define RB_ASSERT_ARG(x) \
+    do {                                \
+        if (RB_UNLIKELY(!(x))) {        \
+            assert(#x && 0);            \
+            return RB_BAD_ARG;          \
+        }                               \
+    } while (0)
+
+/**
+ * The tagged branch is unlikely to be taken
+ */
+#define RB_UNLIKELY(x) __builtin_expect(!!(x), 0)
```

Comment:
Can MSVC handle `__builtin_expect`?
It seems that MSVC does not implement it:
https://visualstudio.uservoice.com/forums/121579-visual-studio-ide/suggestions/16930249-add-something-similar-to-builtin-expect-from-g

Reply:
I do not really much care about Windows compatibility with this commit, since stream-
processor is not compilable on MSVC anyway (This fix is for Linux and BSD)

I'm planning to submit another PR after this patch to cover up Windows. I do
not think though how to do it yet, though.

Reply:
Ah, got it. Thank you for your reply.